arangod/RestHandler/RestReplicationHandler.cpp

```diff
@@ -2700,16 +2589,144 @@ ReplicationApplier* RestReplicationHandler::getApplier(bool& global) {
   }
 }
 
-//////////////////////////////////////////////////////////////////////////////
-/// @brief condition locker to wake up holdReadLockCollection jobs
-//////////////////////////////////////////////////////////////////////////////
+Result RestReplicationHandler::createBlockingTransaction(aql::QueryId id,
+                                                         LogicalCollection& col,
+                                                         double ttl) const {
+  // This is a constant JSON structure for Queries.
+  // We do not need nodes or variables.
+  // We only need collections, with corresponding lock type
+  auto planBuilder = std::make_shared<VPackBuilder>();
+  planBuilder->openObject();
+  planBuilder->close();
+
+  auto query = std::make_unique<aql::Query>(
+    false,
+    _vocbase,
+    planBuilder,
+    nullptr, /* options */
+    aql::QueryPart::PART_MAIN /* Do locking */
+  );
+ // NOTE: The collections are on purpose not locked here.
+  // To accuire an EXCLUSIVE lock may require time under load,
+  // we want to allow to cancel this operation while waiting
+  // for the lock.
 
-arangodb::basics::ConditionVariable RestReplicationHandler::_condVar;
+  auto queryRegistry = QueryRegistryFeature::QUERY_REGISTRY;
+  TRI_ASSERT(queryRegistry != nullptr);
+  if (queryRegistry == nullptr) {
+    return {TRI_ERROR_SHUTTING_DOWN};
+  }
 
-//////////////////////////////////////////////////////////////////////////////
-/// @brief global table of flags to cancel holdReadLockCollection jobs, if
-/// the flag is set of the ID of a job, the job is cancelled
-//////////////////////////////////////////////////////////////////////////////
+  {
+    // we need to lock in EXCLUSIVE mode here, because simply locking
+    // in READ mode will not stop other writers in RocksDB. In order
+    // to stop other writers, we need to fetch the EXCLUSIVE lock
+    AccessMode::Type access = EngineSelectorFeature::ENGINE->typeName() == "rocksdb"
+      ? AccessMode::Type::EXCLUSIVE : AccessMode::Type::READ;
+    auto ctx = transaction::StandaloneContext::Create(_vocbase);
+    auto trx = std::make_unique<SingleCollectionTransaction>(ctx, col, access);
+    query->setTransactionContext(ctx);
+    query->injectTransaction(trx.release());
+  }
+  auto trx = query->trx();
+  TRI_ASSERT(trx != nullptr);
+  trx->addHint(transaction::Hints::Hint::LOCK_ENTIRELY);
+ 
+  TRI_ASSERT(isLockHeld(id).is(TRI_ERROR_HTTP_NOT_FOUND));
+
+  try {
+    queryRegistry->insert(id, query.get(), ttl, true, true);
```

Comment:
mmmh, can this be changed so that `std::move(query)` is passed to QueryRegistry::insert instead of passing the raw pointer and then releasing it somewhere below?